packages/jest-resolve/src/default_resolver.js

```diff
@@ -122,15 +125,25 @@ function resolveSync(target: Path, options: ResolverOptions): Path {
       return undefined;
     }
 
-    const pkgfile = path.join(name, 'package.json');
-    let pkgmain;
+    let pkgentry, isRequiredByModuleField;
     try {
-      const body = fs.readFileSync(pkgfile, 'utf8');
-      pkgmain = JSON.parse(body).main;
+      const pkgfile = getPackageFile(name);
+
+      if (hasModuleField(pkgfile)) {
+        isRequiredByModuleField = true;
+        pkgentry = pkgfile.module;
+      } else {
+        pkgentry = pkgfile.main;
+      }
     } catch (e) {}
 
-    if (pkgmain && pkgmain !== '.') {
-      const resolveTarget = path.resolve(name, pkgmain);
+    if (pkgentry && pkgentry !== '.') {
+      const resolveTarget = path.resolve(name, pkgentry);
+
+      if (isRequiredByModuleField) {
+        REQUIRED_BY_MODULE_FIELD_CACHE[name] = true;
```

Comment:
Sadly I reverted to keeping a local CACHE singleton here, if you have any ideas, please let me know as well!